jscomp/bsb/bsb_build_ui.ml

```diff
@@ -133,137 +133,148 @@ let (++) (u : t)  (v : t)  =
+                         let excludes = !excludes in 
+                         let file_array = Bsb_dir.readdir dir in 
+                         sources := 
+                           Array.fold_left (fun acc name -> 
+                               if Str.string_match re name 0 && 
+                                  not (List.mem name excludes)
+                               then 
+                                 Binary_cache.map_update  ~dir acc name 
+                               else acc
+                             ) String_map.empty file_array;
+                         globbed_dirs :=  [dir]
+                      ))
+                |> ignore
+              )
+           )             
+        |? (Bsb_build_schemas.bs_dependencies, `Arr (fun s -> bs_dependencies := get_list_string s ))
+        |?  (Bsb_build_schemas.resources ,
+             `Arr (fun s  ->
+                 resources := get_list_string s
+               ))
+        |? (Bsb_build_schemas.public, `Str (fun s -> 
+            if s = Bsb_build_schemas.export_all then public := Export_all else 
+            if s = Bsb_build_schemas.export_none then public := Export_none else 
+              failwith ("invalid str for" ^ s )
+          ))
+        |? (Bsb_build_schemas.public, `Arr (fun s -> 
+            public := Export_set (String_set.of_list (get_list_string s ) )
+          ) )
+        |? (Bsb_build_schemas.subdirs, `Id (fun s -> 
+            let res  = parsing_sources current_dir_index !dir s in 
+
+            children :=  res.files ; 
+            children_update_queue := res.intervals;
+            children_globbed_dirs := res.globbed_dirs
+          ))
+        |> ignore 
+      in 
+      {
+        files = 
+          {dir = !dir; 
+           sources = !sources; 
+           resources = !resources;
+           bs_dependencies = !bs_dependencies;
+           public = !public;
+           dir_index = current_dir_index;
+          } 
+          :: !children;
+        intervals = !update_queue @ !children_update_queue ;
+        globbed_dirs = !globbed_dirs @ !children_globbed_dirs;
       } 
-      :: !children;
-    intervals = !update_queue @ !children_update_queue ;
-    globbed_dirs = !globbed_dirs @ !children_globbed_dirs;
-  } 
-
+  | _ -> empty 
+(* and parsing_simple_dir dir_index cwd  dir  : t = 
+   parsing_source dir_index cwd (String_map.singleton Bsb_build_schemas.dir dir)
+*)
 
-let  parsing_sources dir_index cwd (file_groups : Ext_json.t array)  = 
+and  parsing_arr_sources dir_index cwd (file_groups : Ext_json.t array)  = 
   Array.fold_left (fun  origin x ->
-      match x with 
-      | `Obj map ->  
-        parsing_source dir_index cwd map ++ origin
-      | _ -> origin
+      parsing_source dir_index cwd x ++ origin 
     ) empty  file_groups 
 
+and  parsing_sources dir_index cwd (sources : Ext_json.t )  = 
+  match sources with   
+  | `Arr file_groups -> 
+    parsing_arr_sources dir_index cwd file_groups.Ext_json.content
```

Comment:
we can not treat `arr` the same as `str`, otherwise it will allow `arr of arr` cases